### ISSUE TITLE
Evaluations Data

### ISSUE THREAD

**Issue by AdamMcAdamson** [UTDNebula/api-tools#6]
We would like to provide evaluation data as part of our API.

To this effect, we need to:
- [ ] Extend our database schema to include Evalution data
- [ ] Create a Golang model conforming to the schema
- [x] Develop a scraper with [chromedp](https://github.com/chromedp/chromedp) to scrape the data available from [https://coursebook.utdallas.edu/ues-report/](https://coursebook.utdallas.edu/ues-report/)

**Comment by kneevin** [comment]
hey! i'd like to take on this scraper.

**Comment by AdamMcAdamson** [comment]
That sound good to me, @hochladen?

Also, this issue should probably be moved to https://github.com/UTDNebula/api-tools.

**Comment by jpahm** [comment]
I'm open to it; though I can say that this could also be implemented as a part of the existing coursebook scraper (as well as anything else we may need to pull from coursebook)

I'm not entirely opposed to having this as a separate scraper, but I'd say it's a matter of considering if the separation of tasks would be worth the added clutter.

**Comment by AdamMcAdamson** [comment]
Right, I had forgotten that we can pull eval data from the coursebook scraper with the speedup.

**Comment by kneevin** [comment]
so should i try to refactor the current scraper so it can scrape the eval. data as well? or is it ok to have two separate scrapers? i think it'll make more sense to refactor the current scraper it'll just take me a bit more time to figure it out

**Comment by jpahm** [comment]
Adding it to the current scraper would probably be best, though I still have to push it since it's still WIP and stored on my PC locally at the moment. It'll probably be a day or two before I do that, since I'm currently in the middle of moving back home.

**Comment by jpahm** [comment]
Yeah so it's fair to say "a day or two" was a vast underestimate of how long this would take to add; regardless, it should be added soon as part of the existing scraper now that other priorities have been taken care of.

**Comment by jpahm** [comment]
I've completed the scraper component of this work, but there are some concerns regarding IP ratelimits to be addressed. A data model and associated database changes still need to be completed.

**Comment by jpahm** [comment]
> I've completed the scraper component of this work, but there are some concerns regarding IP ratelimits to be addressed. A data model and associated database changes still need to be completed.

Upon further investigation of this, I'm not seeing any immediate great solutions for the IP ratelimit problem -- this problem also occurs with scraping courses, but in a much more manageable fashion. Scraping evals leads to a long IP ratelimit every 30-40 evals or so, which obviously isn't sustainable for scraping en masse.

A solution to this issue that I proposed to @iamwood would be to set up an API endpoint for evals that parses and returns specific evals on-the-fly* rather than parsing them all en-masse. I'll discuss this alongside some other things once the semester starts rolling more.

Any thoughts on this issue are welcome!

*Alongside some sort of caching would be preferred

**Comment by democat3457** [comment]
+1 for caching + evals on-the-fly, I think it's a good compromise.

**Comment by jpahm** [comment]
So, after putting together an on-demand scraping endpoint for evals, it seems like we are now being hindered on this front by evals being locked behind captcha verification. I'm not sure if there's any way to circumvent this, but I'm out of ideas for the time being. As such, I'm going to be putting this issue on hold in favor of prioritization of other tasks.

**Comment by jpahm** [comment]
Closing this for now since getting eval data appears to be a lost cause for the foreseeable future.

**Comment by justinschwerdtfeger** [comment]
I was recently thinking about course evaluation scraping, and found this issue. How exactly are course evaluations locked behind captchas? I did a little bit of investigating, but not particularly in depth. It seems you just need a "PTGSESSID" cookie to access the evals, so how does that differ from other coursebook scraping?

I'd be interested in investigating this further (probably after finals), but I'm interested if anyone else knows anything else about this issue

**Comment by jpahm** [comment]
>It seems you just need a "PTGSESSID" cookie to access the evals, so how does that differ from other coursebook scraping?

You're correct that the same cookie is required. However, unless things have changed since I originally investigated this, scraping the evals in the same manner as the courses isn't feasible because the evals endpoint is much quicker to perform an IP ratelimit and/or captcha depending on the scraping method being used (direct API calls vs automated browser). I recall the ratelimit/captcha kicking in after only a handful of evals, which doesn't bode well for mass scraping.

It's not the same as getting the courses where we can just reset the ratelimit via generating a new cookie, at least last I checked. Might be worth fresh investigation to see if something changed or was overlooked.
